### ISSUE TITLE
Linux Installation with MIDI sound

### ISSUE THREAD

**Issue by cinephos** [frescobaldi/frescobaldi#1537]
Here are instructions for a functioning installation of Frescobaldi with MIDI sound, tested on an Ubuntu 22.04LTS system. 
_Posted here in order to get the attention of the programmers, as well as the community in an effort to improve it.
No compilation of source files is necessary, hence easier for the _user.__
1. Install [Frescobaldi (current version 3.2) ](https://flathub.org/apps/details/org.frescobaldi.Frescobaldi)and [Qsynth (current version 0.9.9)](https://flathub.org/apps/details/org.rncbc.qsynth) from Flatpak. In order to do that you need to [install Flatpak](https://flatpak.org/setup/) beforehand.
2. Install fluidsynth with terminal command `sudo apt-get install fluidsynth`. This step is necessary in order to store the necessary soundfonts in your system.
3. Customize Qsynth: Select Setup. 

- On MIDI tab select: MIDI Driver=alsa-seq, MIDI channels=16, MIDI Bank select mode=gm, MIDI CLient name ID=pid, then tick (select) Auto Connect MIDI Inputs.
- On Audio tab select: Audio driver=Alsa, Sample Format=16bit, Sample rate=44100, Buffer size=512, Buffer count=2, Audio device=(leave this box empty), Audio Channels=2, Audio Groups=1, Polyphony=256
- On Soundfonts tab, select Open and then browse directory `/usr/share/sounds/sf2`. `FluidR3_GM.sf2` and `TimGM6mb.sf2` soundfonts should work correctly. You may select either of them. I prefer the larger first one.

4. Open Frescobaldi, go to Preferences/MIDI settings and select Player Output=Synth Input Port (2:0)

That's it! Now you can play the midi output of your Frescobaldi *.ly file.

**Notes**

- Qsynth saves its setup upon exiting.

- If you wish to investigate further Qsynth setup while playback, have in mind that you need to restart Qsynth. (There is a button for this.)

- If you restart Qsynth, you also need to Refresh MIDI ports on Frescobaldi's Preferences/MIDI settings.

- If your MIDI player is in the middle of the score, then you are going to hear the default instruments assigned to the channel(s). You need to return to the start of the midi file in order to hear the instruments defined in your *.ly file with commands like `\set Staff.midiInstrument = "flute"` or `\set ChordNames.midiInstrument = "percussive organ"` etc.

**Comment by fedelibre** [comment]
The [MIDI output page in the user manual](https://github.com/frescobaldi/frescobaldi/blob/master/frescobaldi_app/userguide/midi_synth.md) is just a stub. It needs to be expanded.

There are also guides in the wiki:
- https://github.com/frescobaldi/frescobaldi/wiki/MIDI-playback-on-Mac-OS-X
- https://github.com/frescobaldi/frescobaldi/wiki/MIDI-playback-on-Linux

**Comment by cinephos** [comment]
More soundfonts can be found in [fluidsynth's wiki](https://github.com/FluidSynth/fluidsynth/wiki/SoundFont) page.

**Comment by reznaeous** [comment]
On Sun, Feb 26, 2023 at 1:37 AM cinephos ***@***.***> wrote:

>
>    1. Install Frescobaldi (current version 3.2)
>    <https://flathub.org/apps/details/org.frescobaldi.Frescobaldi>and Qsynth
>    (current version 0.9.9)
>    <https://flathub.org/apps/details/org.rncbc.qsynth> from Flatpak. In
>    order to do that you need to install Flatpak
>    <https://flatpak.org/setup/> beforehand.
>
>
FWIW I'm on Kubuntu 22.10, and QSynth 0.9.7 from the repositories worked
just fine - no need for Flatpak with this release.


>    1. Install fluidsynth with terminal command sudo apt-get install
>    fluidsynth. This step is necessary in order to store the necessary
>    soundfonts in your system.
>    2. Customize Qsynth: Select Setup.
>
>
>    - On MIDI tab select: MIDI Driver=alsa-seq, MIDI channels=16, MIDI
>    Bank select mode=gm, MIDI CLient name ID=pid, then tick (select) Auto
>    Connect MIDI Inputs.
>    - On Audio tab select: Audio driver=Alsa, Sample Format=16bit, Sample
>    rate=44100, Buffer size=512, Buffer count=2, Audio device=(leave this box
>    empty), Audio Channels=2, Audio Groups=1, Polyphony=256
>
>
I had to use PulseAudio instead of ALSA, but other than that all other
settings worked.


>
>    - On Soundfonts tab, select Open and then browse directory
>    /usr/share/sounds/sf2. FluidR3_GM.sf2 and TimGM6mb.sf2 soundfonts
>    should work correctly. You may select either of them. I prefer the larger
>    first one.
>
>
>    1. Open Frescobaldi, go to Preferences/MIDI settings and select Player
>    Output=Synth Input Port (2:0)
>
> That's it! Now you can play the midi output of your Frescobaldi *.ly file.
>

Yup! Working just fine now! Been using an external program for this, but no
need for that anymore. Thanks for the pointers!

Thanks,
Michael

**Comment by cinephos** [comment]
So, I suppose that flatpak is not necessary and Frescobaldi (downloaded from any repository) can work together with Qsynth (downloaded from any repository). The Qsynth settings are important and the soundfont. I tested a sounfont from [fluidsynth's wiki ](https://github.com/FluidSynth/fluidsynth/wiki/SoundFont) (the one that reminds of some other software) and it works. It can be saved anywhere on your computer. As a result, installation of fluidsynth from terminal may also not be necessary. 
Upon re-confirmation from the programmers that a functioning result can be reproduced again and again, I suggest that they provide this procedure (Qsynth settings, soundfont URL or FTP link, and Notes like the ones in my original post) on their project's [Download page](https://frescobaldi.org/download).

**Comment by papioara** [comment]
First of all: Thank you for this great howto! I agree that flatpak is not necessary. On Linux Mint 21.1 Cinnamon the flatpak freezes right after launch and cannot be configured. I had to kill it and installed a slightly earlier version from the repos. I also installed jackd just in case, although I do not need it. There is an error about jackd right after startup, but when you ignore it and continue anyway, you can do your config as described. MIDI playback works great like this in Frescobaldi. I had Timidity before, but it always wanted a MIDI port refresh after every compilation. Now no longer necessary...

**Comment by fedelibre** [comment]
> On Linux Mint 21.1 Cinnamon the flatpak freezes right after launch and cannot be configured.

Can you please open an issue [here](https://github.com/flathub/org.frescobaldi.Frescobaldi/issues/). Post the output of `flatpak run org.frescobaldi.Frescobaldi`. It's very likely a problem in your computer,  but I'd like to check.

**Comment by papioara** [comment]
It seems that I was not clear enough in my statement about the freezing flatpak. I did not mean the frescobaldi one, but the Qsynth flatpak. 

On launching it, I get errors which all my installed flatpaks throw. But apart from that they all work great:

![image](https://user-images.githubusercontent.com/31970041/221872954-c413520a-c517-4928-b727-900f8dc202c8.png)

And so just ignore them. Qsynth throws this here:

![image](https://user-images.githubusercontent.com/31970041/221875393-5f81871e-f867-490d-b1bf-813a8a7d5d67.png)

The window is there, but the buttons do not react to clicks. It's just dead. I do not understand the second line in the error message. What network connection is this about? The Qsynth from the repos throws different errors (see my previous post), but it works after all.

**Comment by jrd10** [comment]
Thanks for this information. With QSynth, the Frescobladi MIDI player works fine :).

Is there a way/view to follow the notes you hear from the MIDI player?

Thank you for Frescobladi.

**Comment by papioara** [comment]
I do not think that this is possible in Frescobaldi itself. The only software I know for such a purpose is PianoBooster. It is available as an AppImage. Its limitation is, though, that its audience is piano players. As long as you have piano music on two staves, it works great. But if you have something with more staves, e.g. a string quartet score, you can only see and follow along the notes from two staves, while you hear all the parts/notes in the score. It is better than nothing, though, if you want something along these lines. 

In Frescobaldi, the player can only show which beat in bar X you are on.

**Comment by papioara** [comment]
Edit: I have discovered another software which is also capable of displaying all parts of a score and following the notes along:
you can download it here: [https://www.notation.com/Player.php](url). It's called Notation Player. Only downside: it is a Windows program, and you need to have Wine installed to be able to run it. It just did a test drive (Linux Mint 21.3 Cinnamon), and it works great on my machine - tested several scores with it. If you get it to run, this definitely is an option.

**Comment by ksnortum** [comment]
On Fri, Mar 1, 2024 at 12:26 PM papioara ***@***.***> wrote:

> I do not think that this is possible in Frescobaldi itself. The only
> software I know for such a purpose is PianoBooster.
>

I believe MuseScore <https://musescore.org/en> will do this too, but then
you're not using LilyPond any more.  You can import a MIDI file into
MuseScore and watch the notes that way, but I don't know how much help that
would be.

**Comment by jrd10** [comment]
Thank you @ksnortum @papioara 
Yes I use Musescore too. From a .midi file :)
